[RFC] Refactor Neovim to remove the side effects of `K_EVENT`

tarruda opened the pull request:
This PR is my new attempt to remove the effects of the internal `K_EVENT` key code, which is used to represent events in various modes that explicitly declare the ability to do so. While using a special key to signal events is relatively safe(because Vim is a state machine that only deals with keys), it has a couple of disadvantages:
- Some operators can be interrupted
- Mappings are affected

Previously I have attempted to remove this key by always handling events in the most inner place of Neovim loops: the input.c module, which created instability since we can't handle events in every state.

Fixing this issue is a bigger deal now because terminal redrawing is now handled as deferred events: https://github.com/neovim/neovim/commit/47cbbc0b8c864bdb1aaa572f0396642e2c66deae

Close #3382 (and possibly many other issues I can't find right now)

tarruda committed:
test: Add more TUI tests and increase timeout

tarruda committed:
main: Extract `normal_check` from `main_loop`

The new function contains logic that must be executed after handling input in
normal mode and also before the first main loop iteration. Also rename
`main_loop` to `normal_enter` and move it to normal.c

tarruda committed:
normal: Extract most `normal_cmd` logic into two functions

The new functions are `normal_prepare` and `normal_execute` which contain code
executed before and after input is received in normal mode.

tarruda committed:
normal: Fix code style in `normal_prepare` and `normal_execute`

This was done separately to make it easier to follow the changes in the previous
commit.

oni-link commented on src/nvim/normal.c:
```diff
@@ -904,164 +877,347 @@ normal_cmd (
+    if (s->noexmode) {
+      return false;
+    }
+    do_exmode(exmode_active == EXMODE_VIM);
```
If `!s->cmdwin || cmdwin_result == 0` is `true` then after the call `do_exmode()`, `normal_cmd()` is now always called. That was not the case in the old code.

tarruda replied:
:+1:

oni-link commented on src/nvim/normal.c:
```diff
@@ -904,164 +877,347 @@ normal_cmd (
+  // inside a ":g/pat/cmd" command, then the "got_int" needs to abort
+  // the ":g" command.
+  // For ":g/pat/vi" we reset "got_int" when used once.  When used
+  // a second time we go back to Ex mode and abort the ":g" command. */
```
`*/` can be removed at the end of the line.

tarruda replied:
:+1:

oni-link commented on src/nvim/normal.c:
```diff
@@ -904,164 +877,347 @@ normal_cmd (
+    if (s->noexmode && global_busy && !exmode_active
+        && s->previous_got_int) {
+      // Typed two CTRL-C in a row: go back to ex mode as if "Q" was
+      // used and keep "got_int" set, so that it aborts ":g". */
```
`*/` can be removed at the end of the line.

tarruda replied:
:+1:

tarruda committed:
input: Remove CURSORHOLD key

Refactor input.c, normal.c and edit.c to use the K_EVENT special key to trigger
the CURSORHOLD event. In normal and edit mode, K_EVENT is treated as
K_CURSORHOLD, which enables better handling of arbitrary actions in those
states(eg: In normal mode the previous operator counts will be restored).

Also fix a test in vim_spec.lua. The test had a wrong assumption: cmdheight is
only used to determine when the press enter screen will be shown, not to limit
how many lines or control pagination.

tarruda committed:
main: Call `normal_execute` from `normal_enter`

`normal_prepare` is now called by `normal_check` before returning 1(to
continue).

Also remove `input_{enable,disable}_events` calls from `normal_cmd`, which only
exists now as a compatibility function to run normal commands with keys inserted
into the typeahead buffer(We don't want to process events in these cases
anyway).

tarruda committed:
main: Refactor normal_enter to call `os_inchar` directly

This makes it impossible for K_EVENT to interfere with mappings, but it also
disables processing of events while in the middle of a mapping (Though this will
be fixed later as this refactoring progresses).

`may_sync_undo` is now called when K_EVENT is received. This is necessary to
correctly update undo entry lists before executing some action.

tarruda committed:
main: Start modeling Nvim as pushdown automaton

From a very high level point of view, Vim/Nvim can be described as state
machines following these instructions in a loop:

- Read user input
- Peform some action. The action is determined by the current state and can
  switch states.
- Possibly display some feedback to the user.

This is not immediately visible because these instructions spread across dozens
of nested loops and function calls, making it very hard to modify the state
machine(to support more event types, for example).

So far, the approach Nvim has taken to allow more events is this:

- At the very core function that blocks for input, poll for arbitrary events.
- If the event received from the OS is user input, just return it normally to
  the callers.
- If the event is not direct result of user input(possibly a vimscript function
  call coming from a msgpack-rpc socket or a job control callback), return a
  special key code(`K_EVENT`) that is handled by callers where it is safer to
  perform arbitrary actions.

One problem with this approach is that the `K_EVENT` signal is being sent across
multiple states that may be unaware of it. This was partially fixed with the
`input_enable_events`/`input_disable_events` functions, which were added as a
mechanism that the upper layers can use to tell the core input functions that it
is ready to accept `K_EVENT`.

Another problem is that the mapping engine is implemented in getchar.c
which is called from every state, but the mapping engine is not aware of
`K_EVENT` so events can break mappings.

While it is theoretically possible to modify getchar.c to make it aware of
`K_EVENT`, this commit fixes the problem with a different approach: Model Nvim
as a pushdown automaton(https://en.wikipedia.org/wiki/Pushdown_automaton). This
design has many advantages which include:

- Decoupling the event loop from the states reponsible for handling events.
- Better control of state transition with less dependency on global variable
  hacks(eg: 'restart_edit' global variable).
- Easier removal of global variables and function splitting. That is because
  many variables are for state-specific information, and probably ended up being
  global to simplify communication between functions, which we fix by storing
  state-specific information in specialized structures.

The final goal is to let Nvim have a single top-level event loop represented by
the following pseudo-code:

```
while not quitting
  let event = read_event
  current_state(event)
  update_screen()
```

This closely mirrors the state machine description above and makes it easier to
understand, extend and debug the program.

Note that while the pseudo code suggests an explicit stack of states that
doesn't rely on return addresses(as suggested by the principles of
automata-based programming:
https://en.wikipedia.org/wiki/Automata-based_programming), for now we'll use the
call stack as a structure to manage state transitioning as it would be very
difficult to refactor Nvim to use an explicit stack of states, and the benefits
would be small.

While this change may seem like an endless amount of work, it is possible to
do it incrementally as was shown in the previous commits. The general procedure
is:

1- Find a blocking `vgetc()`(or derivatives) call. This call represents an
   implicit state of the program.

2- Split the code before and after the `vgetc()` call into functions that match
   the signature of `state_check_callback` and `state_execute_callback.
   Only `state_execute_callback` is required.

3- Create a `VimState` "subclass" and a initializer function that sets the
   function pointers and performs any other required initialization steps. If
   the state has no local variables, just use `VimState` without subclassing.

4- Instead of calling the original function containing the `vgetc()`,
   initialize a stack-allocated `VimState` subclass, then call `state_enter` to
   begin processing events in the state.

5- The check/execute callbacks can return 1 to continue normally, 0 to break the
   loop or -1 to skip to the next iteration. These callbacks contain code that
   execute before and after the old `vgetc()` call.

The functions created in step 2 may contain other `vgetc()` calls. These
represent implicit sub-states of the current state, but it is fine to remove
them later in smaller steps since we didn't break compatibility with existing
code.

tarruda committed:
normal: Split `normal_check` into multiple functions

Split most code in `normal_check` in:

- `normal_check_stuff_buffer`
- `normal_check_interrupt`
- `normal_check_cursor_moved`
- `normal_check_text_changed`
- `normal_check_folds`
- `normal_redraw`

tarruda committed:
normal: Extract some functions from `normal_execute`

- `normal_handle_special_visual_command`
- `normal_need_aditional_char`
- `normal_get_additional_char`
- `normal_invert_horizontal`

tarruda committed:
normal: Extract `normal_get_command_count` from `normal_execute`

tarruda committed:
normal: Extract `normal_finish_command` from `normal_execute`

tarruda committed:
normal: Extract some functions from `normal_finish_command`

- `normal_need_redraw_mode_message`
- `normal_redraw_mode_message`

oni-link commented:
CursorHold is now retriggered (every `'updatetime'` ms) without pressing a key:

vimrc:
```vim
set updatetime=100

let g:x=0

function! Update()
    let g:x=g:x+1
    echom "CursorHoldI ".g:x
    call jobstart('echo')
endfunction

function! Update2()
    let g:x=g:x+1
    echom "CursorHold ".g:x
    call jobstart('echo') 
endfunction

autocmd CursorHoldI * call Update()
autocmd CursorHold * call Update2()
```
Script used with `nvim -u vimrc`.

tarruda commented:
> CursorHold is now retriggered (every 'updatetime' ms) without pressing a key:

:+1: Fixed

tarruda committed:
edit: Extract local variables from edit() and fix code style

Begin refactoring edit() into a state that can be managed by the `state_enter()`:

- Move local variables into a local InsertState structure
- Fix code style in the entire function.

tarruda committed:
edit: Move most code from `edit()` to `insert_{enter,check,execute}`

Refactor insert mode to use `state_enter` as an event loop:

- Move o_lnum(static variable) outside function
- Move code before the insert mode loop into `insert_enter`
- Move code before `safe_vgetc()` call into `insert_check`
- Move code after `safe_vgetc()` call into `insert_execute`
- Remove doESCkey label and handle insert mode repeating in the `insert_enter`
  function
- Remove do_intr label(this is not the place for platform-specific interrupt
  charts)

tarruda committed:
edit: Extract some functions from `insert_execute`

- `insert_handle_key`: Contains the big insert mode switch statement.
- `insert_do_complete`: Code that used to be in the `docomplete` label.
- `insert_do_cindent`: Code that used to be in the `force_cindent` label.

Also move some code after the switch statement into the beginning of
`insert_check`.

tarruda commented:
@ZyX-I I'm having trouble determining why a certain shada test is failing in this PR, more specifically on commit https://github.com/tarruda/neovim/commit/a4be27040537941bf81ce9c552be21ad1bf7616f (main: Refactor main_loop to call `os_inchar` directly).

The failing test can be run with:

```
TEST_FILTER='correctly ignores case' TEST_FILE=test/functional/shada/shada_spec.lua make test
```

and has the following failed assertion:

```
Failure → test/functional/shada/shada_spec.lua @ 189
ShaDa support code correctly ignores case with shada-r option
test/functional/shada/shada_spec.lua:200: Expected objects to be the same.
Passed in:
(table) {
  [7] = 1
  [8] = 2
  [9] = 1
  [10] = 4
 *[11] = 1 }
Expected:
(table) {
  [7] = 1
  [8] = 2
  [9] = 1
  [10] = 4
 *[11] = 2 }
```

That is, key `11` has a expected value of 2 but the actual value was 1. Since I'm having trouble to quickly parse the relevant shada test, can you explain what is the meaning of key `11`?

tarruda committed:
terminal: Refactor to use `state_enter`

- Create `TerminalState` structure containing data used in terminal mode
- Extract `terminal_execute` from `terminal_enter` and use it with
  `state_enter`.

ZyX-I commented:
@tarruda This key is a number of recorded changelist items in the ShaDa file for the file given as the argument to `find_file`. Basically this means that one change has disappeared from change list for some reason.

bfredl commented:
I get this similar error on master
```
ShaDa support code correctly uses shada-r option
/home/bjorn/dev/neovim/test/functional/shada/shada_spec.lua:177: Expected objects to be the same.
Passed in:
(table) { }
Expected:
(table) {
 *[7] = 1
  [8] = 2
  [9] = 1
  [10] = 4
  [11] = 1 }
```

ZyX-I commented:
@bfredl This is by no means similar error. In your case ShaDa file contains nothing about the tested file *at all*, which is very different from missing changelist entry.

tarruda commented:
> @tarruda This key is a number of recorded changelist items in the ShaDa file for the file given as the argument to find_file. Basically this means that one change has disappeared from change list for some reason.

I'd like to debug this with gdb, can you point me to the section of code that reads the number of changelist items into the shada msgpack entry?

ZyX-I commented:
@tarruda This is lua code, function `find_file`.

ZyX-I commented:
And one ShaDa entry *always* contains one (entry type is 11) or zero (any other entry type) changelist items.

tarruda commented:
> @tarruda This is lua code, function find_file.

I meant that code that reads the number of changelist items into a msgpack entry(before writing to the shada file). I want to follow the code paths before and after the commit that introduces the failure

ZyX-I commented:
There is no msgpack entry that contains a number of changelist item. Code that processes change and jump lists is using `MERGE_JUMPS` macros: both when reading and writing the file.

tarruda commented:
Ok, that is a big macro. Can we turn it into a function so it is easier to debug?

tarruda commented:
Nvm, I see that you use some defines to change the behavior of the macro, it would not be as simple as converting to a function

ZyX-I commented:
@tarruda Not in C. There will be a number of `void*` pointers and a number of similarly generic function pointer types.

tarruda commented:
@ZyX-I nevermind, I just wanted to find the relevant variables and they are `b_changelist`/`b_changelistlen`. This should be enough for me to start debugging

tarruda commented:
Found the problem with the shada test: Now that K_EVENT doesn't come from `vgetc()`, `may_sync_undo` isn't called for K_EVENT. This is required to set `curbuf->b_new_change` and `curbuf->b_u_synced` which are used to [increment](https://github.com/neovim/neovim/blob/3bb266266935e9bf95f4a776b33800b9303d4607/src/nvim/misc1.c#L2049-L2093) `curbuf->b_changedlistlen`(serialized by shada). Calling `may_sync_undo` before the mode-specific execute function fixed the issue.

ZyX-I commented:
It is rather strange that this error was caught by a test that exists merely to check that specific file gets ignored with some `shada-r` option value. I.e. it was not testing something related to undo or changelist, just I wrote it so it counts ShaDa entries and not just checks their presence (which would also be enough for the test purposes).

tarruda commented:
> It is rather strange that this error was caught by a test that exists merely to check that specific file gets ignored with some shada-r option value. I.e. it was not testing something related to undo or changelist, just I wrote it so it counts ShaDa entries and not just checks their presence (which would also be enough for the test purposes).

In a software so complex as this one, it is not unusual for this to happen. When testing feature A, we implicitly test feature B, C, D...

The important is that the shada tests added new coverage for existing code

ZyX-I commented:
I know. But it would be much better if it was caught using specific test for this feature and not implicit test which may be changed so, that it will still serve its purpose, but not catch additional problems.

tarruda committed:
ex_getln: Extract local variables from getcmdline() and fix code style

Begin refactoring getcmdline() into a state that can be managed by the
`state_enter()`:

- Move local variables into a local CommandLineState structure
- Fix code style in the entire function.

tarruda committed:
ex_getln: refactor command line mode to use the `state_enter` loop

Split `getcmdline()` into command_line_{enter,check,execute}`

tarruda committed:
ex_getln: Convert `cmdline_{not_,}changed` labels into functions

tarruda committed:
ex_getln: Extract command_line_handle_key from command_line_execute

tarruda committed:
Start documenting code

tarruda commented:
I still have a few ideas that will greatly simplify Neovim code, similarly to how the changes implemented by this branch did: Generalizing source code structures. But this will have to wait until after 0.1 because this PR is already too big

I'd really appreciate anyone helping me test this branch by using it locally. I'm relatively confident no major bugs were introduced because we have a reasonable test suite and I've been using it since the PR started, but it would certainly help if more people were using it in different environments.

justinmk commented:
Will try it for a couple days.

oni-link commented on src/nvim/README.md:
```diff
@@ -0,0 +1,189 @@
+that are constantly changing. As the code becomes more organized and stable,
+this document will be updated to reflect the changes.
+
+If you are looking module-specific details, it is best to read the source code.
```
... looking module-... -> ... looking for module-...

tarruda replied:
:+1:

justinmk commented:
Although this appears to fix https://github.com/neovim/neovim/issues/3382 (by preventing redraw during operator-pending?), this other case still exists:

- `nvim -u NONE -c 'te top'
- `ctrl-\ ctrl-n`
- `:e<tab>` to show wildmenu. Notice that on the next `top` redraw/event, wildmenu disappears.

tarruda commented:
@justinmk for both problems you reported, I don't think there's much I can do without refactoring screen.c. The real issue is that in some situations, no other information about what is on the screen is kept except for what is already on `ScreenLines` and friends.

For example, when `'wildmenu'` is being shown in command-line mode, screen.c has no idea that it is displaying a wildmenu. It was simply drawn over the status area, which is redraw when `update_screen` is called.

In other words, when you press <kbd>enter</kbd> and `'wildmenu'` is being shown, no function is called to hide the menu and redisplay the status line, it is simply a consequence of `update_screen` resetting status text in the next normal loop iteration.

I suppose I could add an extra global variable that disables status lines redrawing while on command-line, but it could be error prone and cause other bugs. I would rather defer fixing this(which is already present on master) to another PR where screen.c is properly refactored.

tarruda committed:
edit.c: Fix one clint warning

oni-link commented:
The use of `intr_char` inside of a `#ifdef UNIX`was removed in some cases in the refactoring. Do we still need it? Can this global variable be removed?

tarruda commented:
> The use of intr_char inside of a #ifdef UNIXwas removed in some cases in the refactoring. Do we still need it? Can this global variable be removed?

The global variable is not set and always has a value of 0, so I guess it can be removed safely

tarruda commented:
In any case, it's not like we want to have platform-specific code in core functions. This translation should happen in the TUI(if it should happen at all)

justinmk commented:
> The real issue is that in some situations, no other information about what is on the screen is kept except for what is already on ScreenLines and friends.

Understood, though in the case of `nnoremap gl :ls<CR>:buffer<Space>` where the pager is showing with Press-Enter prompt, I believe we do have flags that can already be checked.

For reference, this issue is covered by https://github.com/neovim/neovim/issues/3416
